x-pack/plugins/ml/public/components/job_selector/custom_selection_table/custom_selection_table.js

```diff
@@ -0,0 +1,390 @@
+      } else {
+        // grab all id's
+        currentSelected = currentItems.map((item) => item.id);
+      }
+    }
+
+    onTableChange(currentSelected);
+  }
+
+  function handleChangeItemsPerPage(itemsPerPage) {
+    pager.setItemsPerPage(itemsPerPage);
+    setPagerSettings({
+      ...pagerSettings,
+      itemsPerPage,
+      firstItemIndex: pager.getFirstItemIndex(),
+      lastItemIndex: pager.getLastItemIndex()
+    });
+  }
+
+  function handlePageChange(pageIndex) {
+    pager.goToPageIndex(pageIndex);
+    setPagerSettings({
+      ...pagerSettings,
+      firstItemIndex: pager.getFirstItemIndex(),
+      lastItemIndex: pager.getLastItemIndex()
+    });
+  }
+
+  function handleQueryChange({ query: incomingQuery, error: newError }) {
+    if (newError) {
+      setError(newError);
+    } else {
+      const queriedItems = EuiSearchBar.Query.execute(incomingQuery, items, { defaultFields: filterDefaultFields });
+      setError(null);
+      setCurrentItems(queriedItems);
+      setQuery(incomingQuery);
+    }
+  }
+
+  function isItemSelected(itemId) {
+    return itemIdToSelectedMap[itemId] === true;
+  }
+
+  function areAllItemsSelected() {
+    const indexOfUnselectedItem = currentItems.findIndex(item => !isItemSelected(item.id));
+    return indexOfUnselectedItem === -1;
+  }
+
+  function renderSelectAll(mobile) {
+    const selectAll = i18n.translate('xpack.ml.jobSelector.customTable.selectAllCheckboxLabel', {
+      defaultMessage: 'Select all'
+    });
+
+    return (
+      <EuiCheckbox
+        id="selectAllCheckbox"
+        label={mobile ? selectAll : null}
+        checked={areAllItemsSelected()}
+        onChange={toggleAll}
+        type={mobile ? null : 'inList'}
+      />
+    );
+  }
+
+  function toggleItem(itemId) {
+    // If enforcing singleSelection select incoming and deselect the last selected
+    if (singleSelection) {
+      const lastId = lastSelected[0];
+      // deselect last selected and select incoming id
+      setItemIdToSelectedMap({ ...itemIdToSelectedMap, [lastId]: false, [itemId]: true });
+      handleSingleSelectionTableChange(itemId);
+      setLastSelected([itemId]);
+    } else {
+      const isSelected = !isItemSelected(itemId);
+      setItemIdToSelectedMap({ ...itemIdToSelectedMap, [itemId]: isSelected });
+      handleTableChange({ isSelected, itemId });
+    }
+  }
+
+  function toggleAll() {
+    const allSelected = areAllItemsSelected() || itemIdToSelectedMap.all === true;
+    const newItemIdToSelectedMap = {};
+    currentItems.forEach(item => newItemIdToSelectedMap[item.id] = !allSelected);
+    setItemIdToSelectedMap(newItemIdToSelectedMap);
+    handleTableChange({ isSelected: !allSelected, itemId: 'all' });
+  }
+
+  function onSort(prop) {
+    sortableProperties.sortOn(prop);
+    const sortedItems = sortableProperties.sortItems(currentItems);
+    setCurrentItems(sortedItems);
+    setSortedColumn(prop);
+  }
+
+  function renderHeaderCells() {
```

Comment:
Nice to have: This could become a `<HeaderCells .. />` stateless functional component and moved out of this component.